src/libslic3r/GCode.cpp

```diff
@@ -331,6 +332,15 @@ static std::vector<Vec2d> get_path_of_change_filament(const Print& print)
                 this->path.points.begin() + 1,
                 this->path.points.end()
             );
+            
+            // if the wipe path is not big enough to wipe the full requested distance, reduce the retraction amount proportionally
+            // to avoid extreme retraction lengths over a small distance. The remaining retraction distance will be performed in the
+            // parent function
+            if(wipe_path.length() < wipe_dist){
```

Comment:
Rather than reducing the retraction length, I would suggest the following approach. This way it will impact less.

1. Calculate the proportional retraction length according to the actual wipe length, essentially recalculating the length variable.
2. Force a retraction of the extra amount.
3. Perform the wipe and retract using the newly calculated retraction length to avoid excessively fast retraction speeds.

Reply:
How come to retract before rather than after the wipe move? I thought that moving while retracting is more useful at the start as the nozzle pressure is still high vs the opposite?

To do it the proposed way I’d need to create a new retraction function that takes a retraction length amount and executes it and then continue with the code above to wipe the rest. 

However I would have thought that this is less optimal as:
1. Nozzle pressure is the highest at the start of the wipe move, so running the wipe first is the most useful
2. It overrides the user preference to have 0% retraction before wiping for tool changes and small wipe areas
3. It “dry wipes” when doing a filament change as it will retract 2+mm before even doing the wipe move.

that’s why I’ve done it the other way around - wipe then retract what is left, either from a regular retraction or from the tool change. If the user wants a retraction before the wipe, can set the retract before wipe variable in the extruder preferences.

plus it’s cleaner code wise as the current structure allows this change with minimal code intervention :D

thoughts?